modules/core/include/opencv2/core/cuda.hpp

```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS GpuMat
     Allocator* allocator;
 };
 
+class CV_EXPORTS GpuMatND
+{
+public:
+    class CV_EXPORTS Allocator
+    {
+    public:
+        virtual ~Allocator() {}
+
+        // allocator must fill data, step and refcount fields
+        virtual bool allocate(GpuMatND* mat, int* shape, size_t elemSize) = 0;
+        virtual void free(GpuMatND* mat) = 0;
+    };
+
+    //! default allocator
+    static Allocator* defaultAllocator();
+    static void setDefaultAllocator(Allocator* allocator);
+
+    //! default constructor
+    explicit GpuMatND(Allocator* allocator = defaultAllocator());
+
+    //! constructs GpuMatND of the specified shape and type
+    GpuMatND(int* shape, int type, Allocator* allocator = defaultAllocator());
+
+    //! constructs GpuMatND and fills it with the specified value _s
+    GpuMatND(int* shape, int type, Scalar s, Allocator* allocator = defaultAllocator());
+
+    //! copy constructor
+    GpuMatND(const GpuMatND& m);
+
+    //! constructor for GpuMatND headers pointing to user-allocated data
+    GpuMatND(int* shape, int type, void* data, size_t step = Mat::AUTO_STEP);
+
+    //! builds GpuMatND from host memory (Blocking call)
+    explicit GpuMatND(InputArray arr, Allocator* allocator = defaultAllocator());
```

Comment:
Host memory is `Mat` and not `InputArray` (which is some kind of wrapper over multiple types, but it doesn't store any data).

Reply:
Something similar is used in `GpuMat`, so do I need to make changes there as well?